model/system_metrics.go

```diff
@@ -315,3 +368,148 @@ func (id *SystemMetricsInfo) ID() string {
 
 	return fmt.Sprintf("%x", hash.Sum(nil))
 }
+
+// NewSystemMetricsReader returns an io.Reader for the chunks of a particular metric
+// in a provided bucket.
+func NewSystemMetricsReader(ctx context.Context, bucket pail.Bucket, chunks MetricChunks, batchSize int) *SystemMetricsReader {
+	return &SystemMetricsReader{
+		ctx:       ctx,
+		bucket:    bucket,
+		batchSize: batchSize,
+		chunks:    chunks.Chunks,
+		format:    chunks.Format,
+	}
+}
+
+// SystemMetricsReader is a batched, parallelized io.Reader for system metrics data.
+type SystemMetricsReader struct {
+	ctx         context.Context
+	bucket      pail.Bucket
+	batchSize   int
+	chunks      []string
+	chunkIndex  int
+	readerIndex int
+	readers     map[string]io.ReadCloser
+	leftOver    []byte
+	format      FileDataFormat
+	buffer      []byte
+	exhausted   bool
+}
+
+// Format returns the format of the data being read.
+func (s *SystemMetricsReader) Format() FileDataFormat { return s.format }
+
+// Read fills the buffer p with the next data, returning the number of bytes read.
+// If there are no bytes to read, an EOF error is returned.
+func (s *SystemMetricsReader) Read(p []byte) (int, error) {
+	n := 0
+
+	if s.leftOver != nil {
+		data := s.leftOver
+		s.leftOver = nil
+		n = s.writeToBuffer(data, p, n)
+		if n == len(p) {
+			return n, nil
+		}
+	}
+
+	for s.readerIndex < len(s.chunks) {
+		if s.readerIndex > s.chunkIndex {
+			if err := s.getNextBatch(); err != nil {
+				return n, errors.Wrapf(err, "problem loading data")
+			}
+		}
+		data, err := ioutil.ReadAll(s.readers[s.chunks[s.readerIndex]])
+		if err != nil {
+			return n, errors.Wrapf(err, "problem reading data for chunk %s", s.chunks[s.readerIndex])
+		}
+		s.readerIndex += 1
+
+		n = s.writeToBuffer([]byte(data), p, n)
+		if n == len(p) {
+			return n, nil
+		}
+	}
+
+	catcher := grip.NewBasicCatcher()
+	for _, r := range s.readers {
+		catcher.Add(r.Close())
+	}
+	if catcher.HasErrors() {
+		return n, catcher.Resolve()
+	}
+
+	return n, io.EOF
+}
+
+func (s *SystemMetricsReader) writeToBuffer(data, buffer []byte, n int) int {
+	if len(buffer) == 0 {
+		return 0
+	}
+
+	m := len(data)
+	if n+m > len(buffer) {
+		m = len(buffer) - n
+		s.leftOver = data[m:]
+	}
+	_ = copy(buffer[n:n+m], data[:m])
+
+	return n + m
+}
+
+func (s *SystemMetricsReader) getNextBatch() error {
+	catcher := grip.NewBasicCatcher()
+	for _, r := range s.readers {
+		catcher.Add(r.Close())
+	}
+	if err := catcher.Resolve(); err != nil {
+		return errors.Wrap(err, "problem closing readers")
+	}
+
+	end := s.chunkIndex + s.batchSize
+	if end > len(s.chunks) {
+		end = len(s.chunks)
+	}
+
+	work := make(chan string, end-s.chunkIndex)
+	for _, chunk := range s.chunks[s.chunkIndex:end] {
+		work <- chunk
+	}
+	close(work)
```

Comment:
Can we wrap this part in a function/goroutine or something so that we can defer close?

Reply:
It has to close here, so the channel knows not to expect anything else.